include/hipSYCL/sycl/libkernel/sscp/group_functions.hpp

```diff
@@ -556,74 +554,442 @@ T __acpp_joint_reduce(Group g, Ptr first, Ptr last, T init,
+  }
+  return result;
+}
+
+template <typename T, int N, typename BinaryOperation, class Group>
+HIPSYCL_BUILTIN marray<T, N> __acpp_exclusive_scan_over_group(Group g, marray<T, N> x,
+                                                              BinaryOperation binary_op) {
+  marray<T, N> result;
+  for (int i = 0; i < N; ++i) {
+    result[i] = __acpp_exclusive_scan_over_group(g, x[i], binary_op);
+    __acpp_group_barrier(g);
+  }
+  return result;
+}
 
-template <typename T, typename BinaryOperation>
-HIPSYCL_BUILTIN T __acpp_inclusive_scan_over_group(
-    sub_group g, T x, BinaryOperation binary_op);
+template <class Group, typename V, typename T, typename BinaryOperation>
+HIPSYCL_BUILTIN T __acpp_exclusive_scan_over_group(Group g, V x, T init,
+                                                   BinaryOperation binary_op) {
+  const size_t lid = g.get_local_linear_id();
+  auto identity = sscp_binary_operation_identity<std::decay_t<V>,
+                                                 sscp_binary_operation_v<BinaryOperation>>::get();
+  x = lid == 0 ? binary_op(init, x) : x;
+  __acpp_group_barrier(g);
+  x = __acpp_exclusive_scan_over_group(g, x, binary_op);
+  __acpp_group_barrier(g);
+  if (lid == 0) {
+    x = init;
+  }
+  return x;
+}
 
-template <typename Group, typename V, typename T, typename BinaryOperation,
+template <typename Group, typename InPtr, typename OutPtr, typename BinaryOperation,
           std::enable_if_t<is_group_v<std::decay_t<Group>>, bool> = true>
-HIPSYCL_BUILTIN T __acpp_inclusive_scan_over_group(
-    Group g, V x, T init, BinaryOperation binary_op) {
-  auto scan = __acpp_inclusive_scan_over_group(g, T{x}, binary_op);
-  return binary_op(scan, init);
+HIPSYCL_BUILTIN OutPtr __acpp_joint_exclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
+                                                   BinaryOperation binary_op) {
+  const size_t lid = g.get_local_linear_id();
+  __acpp_joint_inclusive_scan(g, first, last - 1, result + 1, binary_op);
+  __acpp_group_barrier(g);
+  using type = decltype(*first);
+  auto identity = sscp_binary_operation_identity<std::decay_t<type>,
+                                                 sscp_binary_operation_v<BinaryOperation>>::get();
+  if (lid == 0) {
+    result[0] = identity;
+  }
+  __acpp_group_barrier(g);
+
+  return result;
+}
+
+template <typename Group, typename InPtr, typename OutPtr, typename T, typename BinaryOperation,
+          std::enable_if_t<is_group_v<std::decay_t<Group>>, bool> = true>
+HIPSYCL_BUILTIN OutPtr __acpp_joint_exclusive_scan(Group g, InPtr first, InPtr last, OutPtr result,
+                                                   T init, BinaryOperation binary_op) {
+
+  const size_t lrange = g.get_local_range().size();
+  const size_t num_elements = last - first;
+  const size_t lid = g.get_local_linear_id();
+  __acpp_group_barrier(g);
+  if (lid == 0 && num_elements > 0) {
+    first[0] = binary_op(first[0], init);
```

Comment:
The SYCL standard does not say this explicitly, but it would be more consistent with `std::exclusive_scan` if the `[first; last)` range was not modified.

Reply:
Should be fixed in https://github.com/AdaptiveCpp/AdaptiveCpp/pull/1709